[1단계 - 상품 관리 기능] 폴로(이동훈) 미션 제출합니다.

green-kong opened the pull request:
안녕하세요 터틀! 🐢
이번 장바구니 미션을 함께 하게 된 폴로입니다! 🙇‍♂️
잘 부탁드리겠습니다.

이번 미션을 진행하면서 고민했던 부분 아래에 남겨 놓겠습니다.

## ❓Q1 Entity에 setter를 써도 될까요?

이번 미션을 진행하면서 이전에는 사용해보지 않던 `BeanPropertyRowMapper`를 사용해 봤습니다.
그런데 이 rowMapper를 쓰려면은 Entity의 모든 필드에 setter가 있어야 하더라구요..!🥲
그래서 이번 미션에는 entity에 setter를 사용했습니다.

그런데 이전 미션에서부터 entity의 setter가 있으면 안되는가? 에 대한 의문은 계속 있었습니다.🤔
setter 없이 생성자만 이용하게 된다면,
입력받는 값이 다르다면 계속 여러개의 생성자가 생길것으로 생각되고,
생성자 오버로딩을 사용할 때 파라미터의 타입이 겹칠수도 있다고 생각했습니다.

반면, setter를 열어둔다면 entity의 값을 변경할 여지가 생긴다고 생각하여 찜찜하기도 합니다.😭
이 부분에 대하여 터틀의 의견을 묻고싶습니다!🙇‍♂️

green-kong committed:
docs: 기능구현 목록 작성

green-kong committed:
feat : db 설정

green-kong committed:
feat : main / admin 페이지 연결

green-kong committed:
feat : table 작성

green-kong committed:
feat : product entity 생성

green-kong committed:
feat: 상품 저장 method 구현

green-kong committed:
feat: 전체 상품 조회하는 method 구현

green-kong committed:
feat: id를 통해 삭제하는 기능 구현

green-kong committed:
feat: id를 통해 수정하는 기능 구현

green-kong committed:
feat: service layer 상품 저장 및 조회 기능 구현

green-kong committed:
feat: service layer 상품 삭제 기능 구현

green-kong committed:
feat: service layer 상품 수정 기능 구현

green-kong committed:
feat: admin controller 상품 등록&조회 기능 구현

green-kong committed:
feat: admin controller 상품 삭제&수정 기능 구현

green-kong committed:
docs: 기능 구현 목록 수정

green-kong committed:
feat: @valid 이용한 입력값 validate, error handling

green-kong committed:
test: 통합테스트코드 작성

green-kong committed:
fix: 중복된 테이블 생성오류 수정

begaonnuri commented on src/main/java/cart/controller/AdminController.java:
```diff
@@ -0,0 +1,59 @@
+        return ResponseEntity.created(URI.create("/admin/product/" + id)).build();
+    }
+
+    @PatchMapping("/product/{id}")
```
Put의 선택지도 있었을텐데 Patch를 사용해주신 이유가 궁금합니다!

green-kong replied:
아 제가 PUT 과 PATCH를 반대로 이해하고 있었나봐요ㅠㅠㅠ

PUT은 전체 데이터를 수정하는 경우에 사용되고,
PATCH는 일부의 데이터를 수정하는데 사용되는 것으로 알고 있습니다.

지금 같은 경우는 전체의 데이터를 수정하기에 PUT의 사용이 더 적절하다고 생각합니다.
수정하도록 하겠습니다!🙇‍♂️

begaonnuri replied:
일부 / 전체를 수정하는지 말고도 멱등성을 유지하는지나 예전 버전 HTTP와 일관성을 유지하기 위해서 등 다양한 이유로 일부 변경임에도 Put을 사용하는 케이스도 있으니 알아두시면 좋을것같아요!

https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObjectAcl.html#API_PutObjectAcl_RequestSyntax

begaonnuri commented on src/main/java/cart/controller/AdminController.java:
```diff
@@ -0,0 +1,59 @@
+    @PatchMapping("/product/{id}")
+    public ResponseEntity<Void> updateProduct(
+            @PathVariable final Long id,
+            @Valid @RequestBody final ProductRequest productRequest) {
```
생성 요청과 수정 요청에 같은 파라미터를 사용하게되면 둘중 하나가 변경됐을때 모든 수정, 생성 요청을 검토해야할것같아요. 이런 문제를 막으려면 어떻게 변경하면 좋을까요?

green-kong replied:
아! 그렇네요!
단순히 같은 정보를 지닌 요청이 들어온다고 생각하여 같은 dto를 사용하였는데,
같은 dto를 사용하기 때문에 변경에 취약할 수 밖에 없는 구조입니다.
게다가 dto는 변경이 잦은 부분이라 더욱더 불안한 요소가 되어버린듯 합니다.
미처 생각하지 못한 부분이네요.

이 부분을 해결하기 위해선 생성과 수정에 사용되는 dto를 각각 만들어 주면 될것 같습니다.
이 부분 적용해서 수정하도록 하겠습니다..!

여기서 드는 한가지 의문은 각각의 api 마다 매핑 되는 각각의 dto가 필요한 걸까요?🤔

begaonnuri replied:
편리함과 명확함 사이에 트레이드오프가 있을 것 같지만 저는 각각의 DTO를 만드는 것을 선호합니다. SRP에 대해 학습해보시는것도 좋을것같아요.

https://velog.io/@hamdoe/%ED%81%B4%EB%A6%B0-%EC%95%84%ED%82%A4%ED%85%8D%EC%B2%98-7.-SRP-%EB%8B%A8%EC%9D%BC-%EC%B1%85%EC%9E%84-%EC%9B%90%EC%B9%99

begaonnuri commented on src/main/java/cart/controller/AdminControllerAdvice.java:
```diff
@@ -0,0 +1,31 @@
+    @ExceptionHandler
+    public ResponseEntity<ErrorResponse> handleBindException(MethodArgumentNotValidException exception) {
+        final String errorMessage = exception.getBindingResult()
+                .getFieldError()
+                .getDefaultMessage();
```
`getFieldError()`엔 null값이 포함될 수 있어서 예외를 처리하다가 오히려 NPE가 발생할 수 있을 것 같아요. 이 부분을 어떻게 개선할 수 있을까요?

<img width="339" alt="image" src="https://user-images.githubusercontent.com/39271364/234601643-8892f0ad-ff23-44bf-863f-fef59c6ce766.png">

green-kong replied:
예외가 터졌을 때, 동작하는 메서드라고 판단을 했고,
그렇기 때문에 getFieldError() 가 nullable 이지만 null을 반환하지 않겠구나 라고 판단하여 null check를 진행하지 않았습니다.

getFieldError() 까지 변수를 선언하고, null check를 진행해주는 방식으로 변경하는 것이 좋을 듯 합니다.

한가지 궁금한 점은, getDefaultMessage() 역시 nullable이기 때문에, 반환값이 null이 된다면 클라이언트 측에서는 에러 메시지가 null로 표시 될것 같은데요.
이 부분 역시 null check를 해주는 것이 좋을까요?

begaonnuri replied:
이부분은 다시 생각해보니 말씀해주신것처럼 null을 반환하지 않을 것 같아서 굳이 null 체크까진 필요 없을것같네요. 과하다고 생각하시면 반영 안해주셔도 좋습니다!

ErrorResponse에 null인 경우를 방지하기 위해선 기본 메시지를 할당하고 생성자 또는 정적 팩터리에서 null이 아닌 경우만 할당해주는 방법도 있을 것 같아요.

begaonnuri commented on src/main/java/cart/controller/AdminControllerAdvice.java:
```diff
@@ -0,0 +1,31 @@
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    @ExceptionHandler
+    public ResponseEntity<ErrorResponse> handleBindException(MethodArgumentNotValidException exception) {
```
handleUnexpectedException 메서드와 달리 handleBindException는 샤옹자가 잘못된 요청을 계속 보내도 알 수 없을것같아요. 다음 링크를 참고해 적당한 레벨의 로그를 남겨보면 어떨까요?

https://www.tutorialspoint.com/log4j/log4j_logging_levels.htm

green-kong replied:
음.. 이부분도 페어와 함께 고민을 했던 부분중 하나인데요...

handleBindException 같은 경우는 사용자의 잘못된 입력으로 발생한 예외를 핸들링하는 메소드 입니다.

서비스에서 사용자의 잘못된 입력은 정말 무수히도 많이 발생할 수 있다고 생각하는데,
이러한 부분까지 log를 남겨야 할까요???

만약 필요하다고 하면 INFO 레벨로 남기는 것이 가장 적당하다고 생각합니다!
좋은 자료 공유 감사합니다👍👍👍

begaonnuri replied:
> 서비스에서 사용자의 잘못된 입력은 정말 무수히도 많이 발생할 수 있다고 생각하는데,
이러한 부분까지 log를 남겨야 할까요???

일반적으론 웹이나 모바일 앱을 통해서도 1차적인 검증이 이루어지기 때문에 이런 부분은 잘못된 요청이라 판단할 여지가 충분하다고 생각해요. 이런 부분은 warn 로그를 남겨 확인하곤 하지만 지금 구조에선 info도 나쁘지 않아보이네요 👍🏻

begaonnuri commented on src/main/java/cart/dao/JdbcProductDao.java:
```diff
@@ -0,0 +1,68 @@
+@Repository
+public class JdbcProductDao implements ProductDao {
+
+    private final NamedParameterJdbcTemplate jdbcTemplate;
```
지금은 구현체가 1개밖에 없지만 객체를 선언할때는 인터페이스로 선언하는 습관을 갖는게 좋습니다!

https://gona.tistory.com/81

green-kong replied:
감사합니다!
터틀의 말씀대로 현재는 구현체가 하나뿐이다 보니, 구현체에 의존을 하게끔 만들었는데
객체지향적인 설계에서는 벗어나버렸네요😅

begaonnuri commented on src/main/java/cart/dto/ProductRequest.java:
```diff
@@ -0,0 +1,38 @@
+
+public class ProductRequest {
+    @NotBlank(message = "상품의 이름이 입력되지 않았습니다.")
+    @Size(min = 1, max = 20, message = "{min}글자 이상 {max}글자 이하로만 입력가능 합니다.")
```
validation 라이브러리의 애너테이션을 적절하게 잘 활용해주셨네요 👍🏻

begaonnuri commented on src/main/java/cart/service/ProductService.java:
```diff
@@ -0,0 +1,54 @@
+        productDao.update(id, createProduct(request));
+    }
+
+    private Product createProduct(ProductRequest productRequest) {
+        return new Product(
+                productRequest.getName(),
+                productRequest.getPrice(),
+                productRequest.getImgUrl()
+        );
+    }
```
ProductService의 기능이 점점 추가되면 이렇게 매핑하는 부분도 상당 부분을 차지하게 될것 같아요. 매핑의 책임을 적절한 객체로 이동해보면 어떨까요? 어떤 객체가 적당할까요?

green-kong replied:
DTO와 entity를 매핑해줄 수 있는 mapper 클래스를 사용하면 좋을 듯 합니다.
다만. 현재는 매핑 하는 부분이 그렇게 크지 않고, 현재 상태에서 클래스를 추가하는 행위가 불필요하게 관리할 클래스가 늘어나는 것은 아닐까 싶어서 따로 클래스를 분리하지는 않았었습니다.

begaonnuri replied:
mapper 클래스도 사용할 수 있지만 매핑이 단순하다면 `productRequest.toDomain()`과 같은 형태로 DTO에 둘 수도 있을 것 같아요.

이 부분은 사람마다 다르겠지만, 우리가 Service 코드를 읽을때 관심사는 어떻게 매핑되는지보다는 비즈니스 로직이 어떤 흐름으로 진행되는지이기 때문에 저는 이런 부분은 DTO나 mapper로 분리하는것이 더 좋다고 생각하긴 합니다.

begaonnuri commented on src/test/java/cart/dao/JdbcProductDaoTest.java:
```diff
@@ -0,0 +1,109 @@
+        assertAll(
+                () -> assertThat(result.getName()).isEqualTo("치킨"),
+                () -> assertThat(result.getPrice()).isEqualTo(10000),
+                () -> assertThat(result.getImgUrl()).isEqualTo("imgUrl")
```
given에 없는 부분을 검증하면 테스트를 읽거나 유지보수하는 팀원이 컨텍스트 파악을 위해 다른 메서드나 클래스를 탐색하게 될 것 같아요. imgUrl 검증은 제외하면 어떨까요?

green-kong replied:
아..! 그런 부분은 전혀 생각지도 못했네요! 감사합니다.🙇‍♂️

begaonnuri commented on src/test/java/cart/dao/JdbcProductDaoTest.java:
```diff
@@ -0,0 +1,109 @@
+        productDao.delete(id);
+
+        // then
+        assertThat(productDao.findAll()).hasSize(1);
```
productDao에서 id를 통해 조회하는 메서드가 없어서 이렇게 검증해주신것 같아요.

findAll에서 가져온 컬렉션에서 해당 id가 없는지를 통해 더 자세히 검증할 수 있지 않을까요?

green-kong replied:
제가 작성한 방식보다 훨씬 더 좋은 검증방법인 것 같습니다! 😍

begaonnuri reviewed (changes requested):
안녕하세요 폴로!
미션 요구사항을 잘 구현해주셨네요! 몇가지 코멘트 남겼는데 확인해주세요.

begaonnuri commented:
### Entity에 setter를 써도 될까요?

폴로가 어떤 것을 중요시 하냐에 따라 다를것같아요.

다른 팀원(예를 들면 저)과 같이 이 프로젝트를 진행할때 코드의 간결함을 중요시하시나요? 아니면 객체를 불변으로 유지해 사이드 이펙트를 줄이는것을 중요시하시나요? 아니면 둘 다 중요시하시나요?

무엇을 더 우선적인 가치로 두느냐에 따라 폴로가 팀원과 함께 선택하면 되는 부분이라고 생각해요.

- 간결함 중시 -> setter
- 사이드 이펙트를 줄이는 것을 중시 -> 생성자
- 혹은 둘 다 중시 -> 리플렉션 API

각 선택지에 대한 장단점을 인지하고 상황에 맞게 최선의 선택을 해보세요 😊

green-kong committed:
fix: update request method 변경 PATCH -> PUT

green-kong committed:
refa: update와 create dto 분리

green-kong committed:
test: 테스트 코드 실행 전 테이블 초기화

green-kong committed:
fix: handleBindException method null check 추가 및 테스트코드 작성

green-kong committed:
feat: handleBindException에 log 생성기능 추가

green-kong committed:
refa: JdbcProductDao dao 의존 타입 impl -> interface

green-kong committed:
refa: dto entity mapping 기능 클래스 분리

green-kong committed:
test: JdbcProductDaoTest#saveProduct imgUrl 파라미터 추가

green-kong committed:
test: JdbcProductDaoTest#delete 검증 방식 변경

before: collection 의 사이즈를 통한 검증
after: collection에 삭제된 프로덕트의 id와 일치하는 프로덕트가 존재하는지 검증

green-kong committed:
feat: 삭제또는 업데이트 시 존재하지 않는 id를 통해 요청이 들어오면 400을 응답하는 기능구현

green-kong committed:
refa: id 못찾는 경우 던지는 예외 변경

before: IllegalArgumentException
after: ResponseStatusException

green-kong commented:
안녕하세요 터틀!
이번주 테코톡을 진행하느라 리뷰반영이 조금 늦어진 것 같아 죄송합니다.😭

우선 터틀이 남겨주신 부분들 모두 반영해봤습니다.

추가적으로 모든 테스트를 한번에 돌려볼때 서로 다른 테스트간에 영향을 미치는 것을 확인하고,
`@Sql` 어노테이션을 이용해 매번 테스트가 진행될 때 마다, 테이블을 초기화 하도록 변경하였습니다.

또한, 업데이트와 삭제의 경우 존재하지 않는 상품 id로 해당 요청이 전송되는 경우, 40를 응답하도록 변경하였습니다.

잘 부탁드리겠습니다!🙇‍♂️

begaonnuri reviewed (approved):
리뷰에 대해 잘 반영해주셨네요 폴로!
사실 한번에 머지해도 충분할만큼 잘 해주셨지만 코멘트에 대한 답을 듣는 과정이 필요해 Request changes를 한번 드렸었어요.
이번 단계는 여기서 머지하겠습니다. 다음단계도 화이팅입니다 💪🏻

begaonnuri commented:
> @Sql 어노테이션을 이용해 매번 테스트가 진행될 때 마다, 테이블을 초기화 하도록 변경하였습니다.

테이블의 데이터로 충돌된것이라면 `@BeforeEach`에서 데이터를 셋업해주고 `@AfterEach`로 데이터를 클리어하는 방식으로 할 수도 있을 것 같네요. 테스트마다 테이블의 재설정이 필요한 것은 아니니까요 😊

begaonnuri commented:
> @Sql 어노테이션을 이용해 매번 테스트가 진행될 때 마다, 테이블을 초기화 하도록 변경하였습니다.

테이블의 데이터로 충돌된것이라면 `@BeforeEach`에서 데이터를 셋업해주고 `@AfterEach`로 데이터를 클리어하는 방식으로 할 수도 있을 것 같네요. 테스트마다 테이블의 재설정이 필요한 것은 아니니까요 😊
